e3db/client.py

```diff
@@ -1141,3 +1149,173 @@ def get_authorizers(self):
+
+        # pending file write created
+        response_json = response.json()
+        upload_file.file_url = response_json["file_url"]
+        upload_file.record_id = uuid.UUID(response_json["id"])
+        headers = {
+            'Content-Type': 'application/octet-stream',
+            'Content-MD5': upload_file.checksum
+        }
+
+        # Read our encrypted file, upload it to E3DB Large Files data storage
+        with open(encrypted_filename, 'rb') as data:
+            # Don't need e3db_auth, since the file url is a signed url just for this session
+            response = requests.put(url=upload_file.file_url, headers=headers, data=data)
+
+        self.__response_check(response)
+        if response.status_code != 200:
+            raise APIError("File upload status code: {0}, body: {1}".format(response.status_code, response.body))
+
+        # File is uploaded now to storage endpoint, need to confirm with E3DB server
+        # to "COMMIT" the file
+        url = self.__get_url("v1", "storage", "files", str(upload_file.record_id))
+        response = requests.patch(url=url, auth=self.e3db_auth)
+        response_json = response.json()
+        # Delete temporary encrypted file, now it is on the server
+        os.remove(encrypted_filename)
+
+        # Construct File Object to be returned to the user
+        return File(
+            response_json['meta']['file_meta']['checksum'],
+            response_json['meta']['file_meta']['compression'],
+            response_json['meta']['file_meta']['size'],
+            response_json['meta']['writer_id'],
+            response_json['meta']['user_id'],
+            response_json['meta']['type'],
+            file_url=response_json['meta']['file_meta']['file_url'],
+            file_name=response_json['meta']['file_meta']['file_name'],
+            record_id=response_json['meta']['record_id'],
+            created=response_json['meta']['created'],
+            last_modified=response_json['meta']['last_modified'],
+            version=response_json['meta']['version'],
+            plain=response_json['meta']['plain']
+        )
+
+    def read_file(self, record_id, destination_filename):
+        """
+        Retrieve and Encrypted file from the server based on record_id.
+        Decrypt the file, and store the plaintext file in destination_filename.
+
+        Parameters
+        ----------
+        record_id : str
+            ID of the record to retrieve
+
+        destination_filename: str
+            Filename string, including path, to store the plaintext file at.
+
+        Returns
+        -------
+        e3db.File
+            File metadata information
+        """
+
+        # Check if destination file can be written to
+        try:
+            destination_file_handle = open(destination_filename, 'w+')
+        except IOError:
+            destination_file_handle.close()
```

Comment:
This is a check to see if the file is writable, and my attempt to make python clean up file handles. If the File is writable, we need to close the file handle (since we are not using it in the future, it was only used to check if we could write to the file). If the file throws an exception during opening, we also need to close the file handle in the except block. If an exception is thrown, the close() method on 1260 will not run, and the file handle would be left open if we did not close it in the except block. It's not the cleanest, but it should work to prevent file handles being left open in the SDK. 
It breaks down like this: